Geth1.6

joelburget opened the pull request:
This PR still requires a bit more work, but I think the time is right for some feedback.

Design goals (in order of priority):

1. Correctness
  * I aim to stay fairly faithful to the original Quorum implementation. This diff started with naively applying Quorum changes to the Geth 1.6 codebase and only making changes to fix errors and reduce the diff size.
  * Unit tests have been incredibly useful. We should have more of them for Quorum.
  * Integration tests (quorum-tools) have similarly been incredibly useful.
  * My unscientific root cause analysis for errors that have cost me the most time so far:
    - Not-specific-enough types: It's the rule rather than the exception to have two or three `StateDB`s or `Address`es on hand when you need to know the right one to use. Would *you* know that it shouldn't be `st.to().Address()`, but rather `*msg.To()` when they're both valid `Address`es? And by the way `st.from().Address()` and `msg.From()` would also work. This is the kind of thing that typecheckers are made for but will take a human hours to figure out.
    - Getters that also set. Did you realize that the `.to()` in `st.to().Address()` in the last point modifies the state? Whoops.
    - Tricky logic. The transaction pool is complex. Execution is complex. There are way too many places we special case a gas price IMHO. Look for `params.IsQuorum` for some places we have special logic for Quorum.
    - `nil` pointers.
2. Small diff
  * Attempting to make it easy to pull in additional improvements in the future. We should be tracking `go-ethereum` master much more closely so we never have to do this again.
  * (TODO) Explore using the `Engine` interface for Raft rather than some of the special cases we currently have.

Most of the commits aren't interesting (and obviously only look at the last dozen or so that are by me). Two that are:

* https://github.com/jpmorganchase/quorum/commit/d996637fc73b191346cbe6d1879cae75d275dadd - add private state
* https://github.com/jpmorganchase/quorum/commit/a924cb629ca754af4f3f49927d15c8e52f41a9d0 -  fix some tricky errors

Misc:

* I use [EIP 155](https://github.com/ethereum/eips/issues/155) or [158](https://github.com/ethereum/eips/issues/158) depending on the block number (just like Geth)
* There are a ton of places where there's just an extra parameter or return value for the private state. We could maybe work around this by changing the type from `state.StateDB` to `state.QuorumStates` (which doesn't currently exist but would just hold but public and private state). I might play around with a couple ideas to see how well they work.
* This diff is meaningless. Don't look at it. Just look at individual commits by me. Don't try to read the commits comprehensively -- the changes are just too large. I added comments on a few of the commits where I'd like targeted feedback.

zsfelfoldi committed:
core: implement ChainIndexer

joelburget committed:
Add Raft mode.

joelburget committed:
Add .idea to gitignore.

lmars committed:
cmd/swarm: Support using Mainnet for resolving ENS names

Signed-off-by: Lewis Marshall <lewis@lmars.net>

bas-vk committed:
whisper: move ShhClient to its own package

bas-vk committed:
whisper: remove gencodec override for config

bas-vk committed:
whisper: remove obsolete api tests

bas-vk committed:
whisper: fallback to default config if none is specified

bas-vk committed:
whisper: renamed errors

bas-vk committed:
whisper: move flags from whisper package to utils

bas-vk committed:
whisper: use syncmap for dynamic configuration options

zsfelfoldi committed:
les: code refactoring (#14416)

This commit does various code refactorings:

- generalizes and moves the request retrieval/timeout/resend logic out of LesOdr
  (will be used by a subsequent PR)
- reworks the peer management logic so that all services can register with
  peerSet to get notified about added/dropped peers (also gets rid of the ugly
  getAllPeers callback in requestDistributor)
- moves peerSet, LesOdr, requestDistributor and retrieveManager initialization
  out of ProtocolManager because I believe they do not really belong there and the
  whole init process was ugly and ad-hoc

karalabe committed:
Merge pull request #14667 from fjl/swarm-fuse-cleanup

swarm/fuse: simplify externalUnmount, use subtests

bas-vk committed:
whisper: use hexutil.UnmarshalFixedText for topic parsing

holiman committed:
cmd/evm, core/vm: add --nomemory, --nostack to evm (#14617)

MaximilianMeister committed:
README: document new config file option (#14348)

lmars committed:
swarm/test: add integration test for 'swarm up' (#14353)

karalabe committed:
Merge pull request #14657 from markya0616/refactor_clique

consensus/clique: fix typo and don't need to add snapshot into recents again

holiman committed:
core: add testcase for txpool

karalabe committed:
cmd/geth: corrently init gas price for CLI CPU mining

fjl committed:
eth/downloader: separate state sync from queue (#14460)

* eth/downloader: separate state sync from queue

Scheduling of state node downloads hogged the downloader queue lock when
new requests were scheduled. This caused timeouts for other requests.
With this change, state sync is fully independent of all other downloads
and doesn't involve the queue at all.

State sync is started and checked on in processContent. This is slightly
awkward because processContent doesn't have a select loop. Instead, the
queue is closed by an auxiliary goroutine when state sync fails. We
tried several alternatives to this but settled on the current approach
because it's the least amount of change overall.

Handling of the pivot block has changed slightly: the queue previously
prevented import of pivot block receipts before the state of the pivot
block was available. In this commit, the receipt will be imported before
the state. This causes an annoyance where the pivot block is committed
as fast block head even when state downloads fail. Stay tuned for more
updates in this area ;)

* eth/downloader: remove cancelTimeout channel

* eth/downloader: retry state requests on timeout

* eth/downloader: improve comment

* eth/downloader: mark peers idle when state sync is done

* eth/downloader: move pivot block splitting to processContent

This change also ensures that pivot block receipts aren't imported
before the pivot block itself.

* eth/downloader: limit state node retries

* eth/downloader: improve state node error handling and retry check

* eth/downloader: remove maxStateNodeRetries

It fails the sync too much.

* eth/downloader: remove last use of cancelCh in statesync.go

Fixes TestDeliverHeadersHang*Fast and (hopefully)
the weird cancellation behaviour at the end of fast sync.

* eth/downloader: fix leak in runStateSync

* eth/downloader: don't run processFullSyncContent in LightSync mode

* eth/downloader: improve comments

* eth/downloader: fix vet, megacheck

* eth/downloader: remove unrequested tasks anyway

* eth/downloader, trie: various polishes around duplicate items

This commit explicitly tracks duplicate and unexpected state
delieveries done against a trie Sync structure, also adding
there to import info logs.

The commit moves the db batch used to commit trie changes one
level deeper so its flushed after every node insertion. This
is needed to avoid a lot of duplicate retrievals caused by
inconsistencies between Sync internals and database. A better
approach is to track not-yet-written states in trie.Sync and
flush on commit, but I'm focuing on correctness first now.

The commit fixes a regression around pivot block fail count.
The counter previously was reset to 1 if and only if a sync
cycle progressed (inserted at least 1 entry to the database).
The current code reset it already if a node was delivered,
which is not stong enough, because unless it ends up written
to disk, an attacker can just loop and attack ad infinitum.

The commit also fixes a regression around state deliveries
and timeouts. The old downloader tracked if a delivery is
stale (none of the deliveries were requestedt), in which
case it didn't mark the node idle and did not send further
requests, since it signals a past timeout. The current code
did mark it idle even on stale deliveries, which eventually
caused two requests to be in flight at the same time, making
the deliveries always stale and mass duplicating retrievals
between multiple peers.

* eth/downloader: fix state request leak

This commit fixes the hang seen sometimes while doing the state
sync. The cause of the hang was a rare combination of events:
request state data from peer, peer drops and reconnects almost
immediately. This caused a new download task to be assigned to
the peer, overwriting the old one still waiting for a timeout,
which in turned leaked the requests out, never to be retried.
The fix is to ensure that a task assignment moves any pending
one back into the retry queue.

The commit also fixes a regression with peer dropping due to
stalls. The current code considered a peer stalling if they
timed out delivering 1 item. However, the downloader never
requests only one, the minimum is 2 (attempt to fine tune
estimated latency/bandwidth). The fix is simply to drop if
a timeout is detected at 2 items.

Apart from the above bugfixes, the commit contains some code
polishes I made while debugging the hang.

* core, eth, trie: support batched trie sync db writes

* trie: rename SyncMemCache to syncMemBatch

karalabe committed:
Merge pull request #14677 from karalabe/miner-cli-gasprice

cmd/geth: corrently init gas price for CLI CPU mining

karalabe committed:
core: ensure transactions correctly drop on pool limiting

karalabe committed:
Merge pull request #14673 from holiman/txfix

core: add testcase for txpool

fjl committed:
cmd/swarm: disable TestCLISwarmUp because it's flaky

fjl committed:
travis.yml: add fakeroot to launchpad builder

markya0616 committed:
core: remove unused events

karalabe committed:
eth: gracefully error if database cannot be opened

bas-vk committed:
whisper: renamed Info#Message to Info#Messages

karalabe committed:
Merge pull request #14681 from fjl/build-fixup

travis.yml, cmd/swarm: fix Travis CI build

fjl committed:
node: don't return non-nil database on error

karalabe committed:
consensus/clique: minor cleanups

karalabe committed:
Merge pull request #14685 from karalabe/ethdb-metrics-fail-fix

eth: gracefully error if database cannot be opened

karalabe committed:
Merge pull request #14596 from markya0616/valid_clique_vote

consensus/clique: choose valid votes

fjl committed:
common/hexutil: wrap errors in json.UnmarshalTypeError

This adds type and struct field context to error messages.
Instead of "hex string of odd length" users will now see "json: cannot
unmarshal hex string of odd length into Go struct field SendTxArgs.from
of type common.Address".

fjl committed:
params: 1.6.6 stable

fjl committed:
params, VERSION: 1.6.7 unstable

karalabe committed:
cmd/puppeth: fix key reuse during faucet deploys

Commit:
swarm/storage: remove panic on invalid chunk

Commit:
Update README.md (#14701)

README: change heading to "Go Ethereum"

Commit:
Makefile: add make swarm command (#14698)

* Makefile: add make swarm command

* Makefile: minor code formatting polishes

karalabe committed:
Merge pull request #14697 from homotopycolimit/master

swarm/storage: remove panic on invalid chunk

bloonfield committed:
rpc: fix closure problem in batch processing (#14688)

Demo of the issue: https://play.golang.org/p/EeTLFfppqC

karalabe committed:
Merge pull request #14687 from markya0616/unused_events

core: remove unused events

karalabe committed:
cmd/geth: fix whisper flag group capitalization

karalabe committed:
Merge pull request #14540 from bas-vk/whisper-api

whisperv5: integrate whisper and implement API

karalabe committed:
Merge pull request #14690 from karalabe/faucet-key-reuse

cmd/puppeth: fix key reuse during faucet deploys

ayeowch committed:
build: fix devel golang detection on debian/ubuntu (#14711)

VoR0220 committed:
accounts/abi: reorganizing package with small fixes (#14610)

* accounts/abi: reorganizing package and some notes and a quick correction of name.

Signed-off-by: RJ Catalano <rj@monax.io>

get rid of some imports

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: move file names

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: fix boolean decode function

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: fix for the array set and for creating a bool

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: be very very very correct

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: fix up error message and variable names

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: take out unnecessary argument in pack method

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: add bool unpack test and add a panic to readBool function

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: fix panic message

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: change from panic to basic error

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: fix nil to false

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: fill out type regex tests and fill with the correct type for integers

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: move packNumbers into pack.go.

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: separation of the testing suite into appropriately named files.

Signed-off-by: RJ Catalano <rj@monax.io>

* account/abi: change to hex string tests.

Signed-off-by: RJ Catalano <rj@monax.io>

* account/abi: fix up rest of tests to hex

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: declare bool at the package level

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: use errors package in the error file.

Signed-off-by: RJ Catalano <rj@monax.io>

* accounts/abi: fix ugly hack and fix error type declaration.

Signed-off-by: RJ Catalano <rj@monax.io>

karalabe committed:
Merge pull request #14686 from fjl/hexutil-json-type-error

common/hexutil: wrap errors in json.UnmarshalTypeError

fjl committed:
core/state: access trie through Database interface, track errors (#14589)

With this commit, core/state's access to the underlying key/value database is
mediated through an interface. Database errors are tracked in StateDB and
returned by CommitTo or the new Error method.

Motivation for this change: We can remove the light client's duplicated copy of
core/state. The light client now supports node iteration, so tracing and storage
enumeration can work with the light client (not implemented in this commit).

Arachnid committed:
eth, les, light: Refactor downloader to use blockchain interface

holiman committed:
core/vm: fix overflow in gas calculation formula

holiman committed:
core/vm : fix testcase for gas calculation

holiman committed:
core/vm: add benchmarks for some ops and precompiles (#14641)

karalabe committed:
Merge pull request #14718 from holiman/gascalc_fix

core/vm: fix overflow in gas calculation formula

Arachnid committed:
eth, les: Refactor downloader peer to use structs

joelburget committed:
Add private state.

karalabe committed:
consensus, core: EIP 100 polishes, fix chain maker diff

This PR polishes the EIP 100 difficulty adjustment algorithm
to match the same mechanisms as the Homestead was implemented
to keep the code uniform. It also avoids a few memory allocs
by reusing big1 and big2, pulling it out of the common package
and into ethash.

The commit also fixes chain maker to forward the uncle hash
when creating a simulated chain (it wasn't needed until now
so we just skipped a copy there).

holiman committed:
core: fix an off-by-one when the block import counts blocks

karalabe committed:
Merge pull request #14727 from holiman/count_txs_right

Fix error when reporting numer of txs in imported blocks

karalabe committed:
Merge pull request #14646 from ethersphere/swarm-ens-api

cmd/swarm: Support using Mainnet for resolving ENS names

lmars committed:
cmd/swarm: Exit if --ethapi is set

The previous attempt to use --ethapi as a fallback if --ens-api is not
set does not work because --ens-api has a default value, and also
setting --ens-api to "" is the suggested way to disable ENS lookups.

Signed-off-by: Lewis Marshall <lewis@lmars.net>

karalabe committed:
Merge pull request #14732 from ethersphere/swarm-remove-ethapi

cmd/swarm: Exit if --ethapi is set

obscuren committed:
consensus/ethash: implement Metropolis EIP 100

holiman committed:
core: Change local-handling to use sender-account instead of tx hashes

holiman committed:
core: Prevent local tx:s from being discarded

Arachnid committed:
eth/downloader, les, light: Changes in response to review

karalabe committed:
params: remove redundant consts, disable metro on AllProtocolChanges

karalabe committed:
Merge pull request #14749 from karalabe/disable-metro-allprotocols

params: remove redundant consts, disable metro on AllProtocolChanges

Arachnid committed:
eth/downloader: Doc fixes

karalabe committed:
core, eth, les: polish txpool API around local/remote txs

karalabe committed:
eth: drop leftover from previous nonce protection scheme

karalabe committed:
cmd, core: add --txpool.nolocals to disable local price exemptions

karalabe committed:
core: handle nolocals during add, exepmt locals from expiration

ligi committed:
core: fix typo in error message (#14763)

karalabe committed:
core: test locals support in txpool queue limits, fix

The commit reworks the transaction pool queue limitation tests
to cater for testing local accounts, also testing the nolocal flag.

In addition, it also fixes a panic if local transactions exceeded
the global queue allowance (no accounts left to drop from) and also
fixes queue eviction to operate on all accounts, not just the one
being updated.

Arachnid committed:
Merge pull request #14723 from Arachnid/downloadrefactor

Refactor downloader to use interfaces instead of callbacks

eminogrande committed:
README: add missing full stop (#14766)

karalabe committed:
Merge pull request #14737 from holiman/txpool_localaccounts

Txpool localaccounts

Commit:
internal/web3ext: fix debug.traceBlockFromFile wrapper (#14774)

As stated in the documentation, this method should be called traceBlockFromFile
and not traceBlockByFile. Previously this would result in a 'The method ... does
not exist/is not available' error.

karalabe committed:
vendor: update go-stack to fix a sigpanic panic (#14790)

fjl committed:
tests: update tests and implement general state tests (#14734)

Tests are now included as a submodule. This should make updating easier
and removes ~60MB of JSON data from the working copy.

State tests are replaced by General State Tests, which run the same test
with multiple fork configurations.

With the new test runner, consensus tests are run as subtests by walking
json files. Many hex issues have been fixed upstream since the last
update and most custom parsing code is replaced by existing JSON hex
types. Tests can now be marked as 'expected failures', ensuring that
fixes for those tests will trigger an update to test configuration. The
new test runner also supports parallel execution and the -short flag.

karalabe committed:
dockerignore: ignore all git metadata and all tests

karalabe committed:
Merge pull request #14792 from karalabe/slim-docker-image

dockerignore: ignore all git metadata and all tests

fjl committed:
params: v1.6.7 stable

fjl committed:
params, VERSION: v1.6.8 unstable

Dexaran committed:
Update README.md

Dexaran committed:
Update README.md

Dexaran committed:
Update README.md

Dexaran committed:
Update README.md

Dexaran committed:
Update README.md

karalabe committed:
VERSION, params: begin 1.7.0 cycle (cannot downgrade)

karalabe committed:
core: remove redundant storage of transactions and receipts (#14801)

* core: remove redundant storage of transactions and receipts

* core, eth, internal: new transaction schema usage polishes

* eth: implement upgrade mechanism for db deduplication

* core, eth: drop old sequential key db upgrader

* eth: close last iterator on successful db upgrage

* core: prefix the lookup entries to make their purpose clearer

mcdee committed:
common: Address.Hex() outputs EIP55-compliant string

mcdee committed:
common: tests for EIP55-compliant Address.Hex()

mcdee committed:
cmd/geth: update tests for EIP55-compliant Address.Hex()

holiman committed:
core: blocknumber in genesis as hex (#14812)

karalabe committed:
core, ethclient: implement Metropolis EIP 98 (#14750)

Implements ethereum/EIPs#98

karalabe committed:
Merge pull request #14733 from karalabe/metro-eip100

consensus/ethash, core: implement Metropolis EIP 100

karalabe committed:
README: get rid of the on-existent disasm command

karalabe committed:
Merge pull request #14807 from Dexaran/master

README: get rid of the non-existent disasm command

ligi committed:
mobile: use EIP155 signer (#14817)

* mobile: Use EIP155Signer - closes #14762

* mobile: Correctly fall back on HomesteadSigner when no chainID is passed in

Commit:
mobile: don't retain transient []byte in CallMsg.SetData (#14804)

* mobile: don't retain transient []byte in CallMsg.SetData

Go mobile doesn't copy []byte parameters, for performance and to allow
writes to the byte array be reflected in the native byte array.
Unfortunately, that means []byte arguments are only valid during the
call it is being passed into.

CallMsg.SetData retains such a byte array. Copy it instead

Fixes #14675

* mobile: copy all []byte arguments from gomobile

To avoid subtle errors when accidentially retaining an otherwise
transient byte slice coming from gomobile, copy all byte slices before
use.

* mobile: replace copySlice with common.CopyBytes

karalabe committed:
cmd/puppeth: limit cotnainers to 10MB logs

karalabe committed:
cmd/faucet: ignore whitespace in gist content

karalabe committed:
Merge pull request #14823 from karalabe/puppeth-limited-logging

cmd/puppeth: limit cotnainers to 10MB logs

karalabe committed:
Merge pull request #14824 from karalabe/faucet-ignore-whitespace

cmd/faucet: ignore whitespace in gist content

MysticRyuujin committed:
node: Rename TrusterNodes (#14827)

* node: Rename TrusterNodes

* node: Rename TrusterNodes

fjl committed:
core/vm: remove logging and add section labels to struct logs (#14782)

leos committed:
cmd/abigen: update generated go file header text (#14845)

As per https://golang.org/s/generatedcode. This will allow other tools
such as golint to properly ignore the files.

lmars committed:
swarm/api/http: redirect root manifest requests to include trailing slash (#14806)

xmikus01 committed:
trie: typo in comment

cdetrio committed:
core/vm/runtime: fix evm command to use --gasprice flag value

karalabe committed:
Merge pull request #14859 from cdetrio/evm-flag-gasprice

core/vm/runtime: fix evm command to use --gasprice flag value

karalabe committed:
Merge pull request #14858 from xmikus01/patch-1

trie: typo in comment

karalabe committed:
cmd, core, eth: journal local transactions to disk (#14784)

* core: reduce txpool event loop goroutines and sync structs

* cmd, core, eth: journal local transactions to disk

* core: journal replacement pending transactions too

* core: separate transaction journal from pool

am2rican5 committed:
mobile: add RinkebyGenesis method

karalabe committed:
Merge pull request #14882 from am2rican5/rinkeby-genesis

mobile: add RinkebyGenesis method

markya0616 committed:
core: avoid write existing block again (#14849)

Commit:
Makefile: fixed GOBIN absolute path issue (#14854)

bas-vk committed:
light: update txpool signer to EIP155 (#14720)

fjl committed:
common: EIP55-compliant Address.Hex() (#14815)

This patch updates the Address type in common/types.go so that the Hex
function provides an EIP55-compliant output string. The implementation is pretty lightweight; 
on my laptop the benchmark gives 1100ns/op, with the majority of that value due to the Keccak hash.

Commit:
core/asm: fix hex number lexing (#14861)

lmars committed:
cmd/swarm: support exporting, importing chunk db (#14868)

lmars committed:
swarm/api: make api.NewManifest synchronous (#14880)

Previously, NewManifest was asynchronous so subsequent code which tried
to use the returned manifest could error as the manifest was not yet
persisted.

mcdee committed:
ethclient: add NetworkID method (#14791)

There is currently no simple way to obtain the network ID from a Client. 
This adds a NetworkID method that wraps the net_version JSON-RPC call.

gsalgado committed:
Fix TestPrivateTransaction to actually create private transactions (#144)

It now creates private transactions, but that means it now needs to run
constellation in a subprocess, so do that as well

karalabe committed:
accounts, console, internal: support trezor hardware wallet

karalabe committed:
cmd: add makecache cmd, use caches during import cmd

Commit:
Makefile, README: remove evm target, add puppeth to table (#14886)

karalabe committed:
core: polish chain indexer a bit

karalabe committed:
Merge pull request #14889 from karalabe/ethash-cache-on-import

cmd: add makecache cmd, use caches during import cmd

evgk committed:
core/vm: fix typo in comment (#14894)

karalabe committed:
cmd/puppeth: remove wrapping loop in single reads

detailyang committed:
Makefile: call shell function to get pwd

karalabe committed:
Merge pull request #14898 from detailyang/patch-1

Makefile: call shell function to get pwd

karalabe committed:
Merge pull request #14897 from karalabe/cardinal-sin

cmd/puppeth: remove wrapping loop in single reads

egonelbre committed:
common: fix megacheck warnings

egonelbre committed:
consensus: fix megacheck warnings

egonelbre committed:
console: fix megacheck warnings

0xaguspunk committed:
node: fix doc typo

egonelbre committed:
eth: fix megacheck warnings

egonelbre committed:
ethstats: fix megacheck warnings

egonelbre committed:
log: fix megacheck warnings

egonelbre committed:
miner: fix megacheck warnings

egonelbre committed:
node: fix megacheck warnings

karalabe committed:
Merge pull request #14905 from armellini13/patch-1

node: fix doc typo

karalabe committed:
build: drop yakkety builds (launchpad end of life)

karalabe committed:
Merge pull request #14910 from karalabe/drop-yakkety-support

build: drop yakkety builds (launchpad end of life)

karalabe committed:
core: bump timeout test to avoid flakyness on overloaded ci

karalabe committed:
Merge pull request #14911 from karalabe/txpool-flaky-test

core: bump timeout test to avoid flakyness on overloaded ci

egonelbre committed:
accounts: fix megacheck warnings (#14903)

* accounts: fix megacheck warnings

* accounts: don't modify abi in favor of full cleanup

karalabe committed:
Merge pull request #14913 from egonelbre/megacheck_common

common: fix megacheck warnings

karalabe committed:
Merge pull request #14915 from egonelbre/megacheck_console

console: fix megacheck warnings

karalabe committed:
Merge pull request #14921 from egonelbre/megacheck_log

log: fix megacheck warnings

karalabe committed:
Merge pull request #14918 from egonelbre/megacheck_ethstats

ethstats: fix megacheck warnings

karalabe committed:
Merge pull request #14923 from egonelbre/megacheck_node

node: fix megacheck warnings

karalabe committed:
Merge pull request #14922 from egonelbre/megacheck_miner

miner: fix megacheck warnings

egonelbre committed:
consensus: revert time.Until change

egonelbre committed:
cmd: fix megacheck warnings (#14912)

* cmd: fix megacheck warnings

* cmd: revert time.Until changes, keep readFloat

karalabe committed:
Merge pull request #14914 from egonelbre/megacheck_consensus

consensus: fix megacheck warnings

karalabe committed:
core: fix blockchain goroutine leaks in tests

fjl committed:
internal/build: add GoTool and document why it uses GOROOT

karalabe committed:
Merge pull request #14927 from karalabe/blockchain-test-cleanups

core: fix blockchain goroutine leaks in tests

karalabe committed:
Merge pull request #14928 from fjl/build-goroot-explain

internal/build: add GoTool and document why it uses GOROOT

egonelbre committed:
internal: fix megacheck warnings (#14919)

egonelbre committed:
light: fix megacheck warnings (#14920)

karalabe committed:
Merge pull request #14933 from egonelbre/megacheck_eth

eth: fix megacheck warnings

karalabe committed:
core: fix txpool journal and test races

egonelbre committed:
rpc: fix megacheck warnings

egonelbre committed:
ethdb: add basic and parallel sanity tests (#14938)

* ethdb: add basic sanity test

* ethdb: test MemDatabase

* ethdb: add parallel tests

karalabe committed:
Merge pull request #14940 from karalabe/txpool-races

core: fix txpool journal and test races

egonelbre committed:
crypto: fix megacheck warnings (#14917)

* crypto: fix megacheck warnings

* crypto/ecies: remove ASN.1 support

fjl committed:
Merge pull request #14522 from ethereum/go-ethereum/chainproc2

egonelbre committed:
whisper: fix megacheck warnings (#14925)

* whisper: fix megacheck warnings

* whisper/whisperv5: regenerate json codec to fix unused override type

zsfelfoldi committed:
les: fix megacheck warnings (#14941)

* les: fix megacheck warnings

* les: fixed testGetProofs

egonelbre committed:
contracts: fix megacheck errors (#14916)

* contracts: fix megacheck errors

* contracts: drop useless sleep, lets see what breaks

divan committed:
eth/downloader: fix race in downloadTesterPeer (#14942)

* eth/downloader: fix race in downloadTesterPeer

Signed-off-by: Ivan Daniluk <ivan.daniluk@gmail.com>

* eth/downloader: minor datarace fix cleanup

egonelbre committed:
swarm: fix megacheck warnings

karalabe committed:
accounts: refactor API for generalized USB wallets

fjl committed:
tests: update tests, use blockchain test "network" field

Blockchain tests now include the "network" which determines the chain
config to use. Remove config matching based on test name and share the
name-to-config index with state tests.

Byzantium/Constantinople tests are still skipped because most of them
fail anyway.

obscuren committed:
core: add Metropolis pre-compiles (EIP 197, 198 and 213)

karalabe committed:
core/vm: polish precompile contract code, add tests and benches

 * Update modexp gas calculation to new version
 * Fix modexp modulo 0 special case to return zero

rjl493456442 committed:
ethdb: return copied value from MemDatabase.Get (#14958)

MaximilianMeister committed:
cmd/geth, cmd/swarm: sort commands and flags by name (#3462)

joelburget committed:
core/vm/runtime: remove unused state parameter to NewEnv (#14953)

* core: Remove unused `state` parameter to `NewEnv`.

`cfg.State` is used instead.

* core/vm/runtime: remove unused import

karalabe committed:
Merge pull request #14964 from fjl/tests-update-2

tests: update tests, use blockchain test "network" field

karalabe committed:
console: use keypad based pinpad (Trezor request)

holiman committed:
core/vm: benchmarking of metro precompiles

karalabe committed:
core/vm: optimize copy-less data retrievals

karalabe committed:
Merge pull request #14885 from karalabe/trezor-boom

accounts, console, internal: support trezor hardware wallet

karalabe committed:
Merge pull request #14951 from egonelbre/megacheck_swarm

swarm: fix megacheck warnings

karalabe committed:
Merge pull request #14959 from karalabe/metropolis-precompiles

core/vm: metropolis precompiles

obscuren committed:
core/vm: implement metropolis static call opcode

holiman committed:
cmd/evm: add --receiver, support code from stdin (#14873)

karalabe committed:
core/vm: minor polishes, fix STATICCALL for precompiles

 * Fix STATICCALL so it is able to call precompiles too
 * Fix write detection to use the correct value argument of CALL
 * Fix write protection to ignore the value in CALLCODE

patrickmn commented on HACKING.md:
```diff
@@ -0,0 +1,89 @@
+
+Replace with appropriate absolute paths:
+
+TODO(joel): figure out how to use relative paths
```
Relative path should work just fine for Constellation config and PRIVATE_CONFIG env var. What's the issue?

karalabe committed:
Merge pull request #14978 from karalabe/metropolis-staticcall

core/vm: implement metropolis static call opcode

obscuren committed:
core/vm: implement RETURNDATA metropolis opcodes

karalabe committed:
core/vm: polish RETURNDATA, add missing returns to CALL*

karalabe committed:
tests: pull in latest tests from upstream

karalabe committed:
Merge pull request #14981 from karalabe/metropolis-returndata

core/vm: implement RETURNDATA metropolis opcodes

obscuren committed:
core/vm: implement REVERT metropolis opcode

karalabe committed:
core/vm: rework reversion to work on a higher level

karalabe committed:
tests: skip the bad tests from colliding account touches

joelburget committed:
Squashed commits.

* Remove old note about testing with constellation.
* Review: forceParseRfc3339 -> mustParseRfc3339
* Add isQuorum parameter to SignTx.
* Re-enable TestLogReorgs.
* Fix `SignTx` calls in executables other than geth.
* Bring vm closer to public ethereum implementation.
* Restore all original Ethereum tests.
* Update for new event interface.
* Tweak core / EVM.
* Improve "TX failed" error message.
* Fix "err creating contract" bug.
* Don't log "invalid mix digest" in Quorum.
* Fix public / private bug.
  A private contract calling a method on a public contract would create a
  ghost of the private contract on public state due to a value transfer of
  0.
* Clearer naming in minter / work.commitTransactions.
* Remove unused log import.
* Remove unused `enableQuorumChecks` flag.

patrickmn commented on accounts/keystore/keystore.go:
```diff
@@ -0,0 +1,493 @@
+		return nil, ErrLocked
+	}
+	// Depending on the presence of the chain ID, sign with EIP155 or homestead
+	if chainID != nil { // && !params.IsQuorum {
```
Add clarification of thinking?

joelburget replied:
I hadn't completely thought through the details. This change requires threading through a bit of state and tests were passing. I just created a [branch](https://github.com/joelburget/go-ethereum/commit/f924a27d67ab2eb4d0b2dede278fec1bd832de28) that handles this properly. I think it looks good. Thoughts?

patrickmn commented on consensus/ethash/consensus.go:
```diff
@@ -0,0 +1,572 @@
+	errInvalidPoW        = errors.New("invalid proof-of-work")
+)
+
+func forceParseRfc3339(str string) time.Time {
```
Idiomatic naming is 'mustParseRfc3339'

joelburget replied:
fixed

patrickmn commented on core/block_validator.go:
```diff
@@ -19,337 +19,129 @@ package core
-			branch := fmt.Sprintf("  O - %x\n  |\n", block.Hash())
-			for h := range ancestors {
-				branch += fmt.Sprintf("  O - %x\n  |\n", h)
+	// TODO(joel)
```
What will happen with this?

joelburget replied:
I'll re-enable this tomorrow morning.

patrickmn commented on core/blockchain.go:
```diff
@@ -752,241 +816,278 @@ func (self *BlockChain) InsertReceiptChain(blockChain types.Blocks, receiptChain
+			parent = chain[i-1]
 		}
+
+		// alias state.New because we introduce a variable named state on the next line
```
This shadowing smells. Maybe call state 'st'?

joelburget replied:
It totally smells. But I think it's more important to keep the diff small between Quorum and Geth. But I could change the name if others also want.

patrickmn replied:
Oh. I missed that. Nevermind.

patrickmn commented on core/blockchain_test.go:
```diff
@@ -965,34 +852,34 @@ func TestChainTxReorgs(t *testing.T) {
 	}
 }
 
+/*
```
Clarify why disabled?

joelburget replied:
Re-enabled

patrickmn commented on core/vm/evm.go:
```diff
@@ -0,0 +1,426 @@
+	readOnlyDepth     uint
+}
+
+// NewEVM returns a new EVM evmironment. The returned EVM is not thread safe
```
evm-ironment :)

joelburget replied:
I can't take credit for this :)

patrickmn reviewed (commented):
Partial (mostly style) review

karalabe committed:
core/vm, crypto/bn256: fix bn256 use and pairing corner case

karalabe committed:
Merge pull request #14993 from karalabe/bn256-precompile-fixes

core/vm, crypto/bn256: fix bn256 use and pairing corner case

markya0616 committed:
eth: send but not announce block to peers if propagate is true

karalabe committed:
cmd/puppeth: support blacklisting malicious IPs on ethstats

karalabe committed:
Merge pull request #14999 from karalabe/puppeth-ethstats-blacklist

cmd/puppeth: support blacklisting malicious IPs on ethstats

karalabe committed:
Merge pull request #14996 from markya0616/send_not_announce

eth: send but not announce block to peers if propagate is true

fjl committed:
node: fix instance dir locking and improve error message

The lock file was ineffective because opening leveldb storage in
read-only mode doesn't really take the lock. Fix it by including a
dedicated flock library (which is actually split out of goleveldb).

tailingchen committed:
core, light: send chain events using event.Feed (#14865)

rjl493456442 committed:
core: implement Metropolis EIP 658, receipt status byte

baumstern committed:
Adding period at end of sentence

missing period at line 119

karalabe committed:
Merge pull request #15016 from gurrpi/patch-1

README: add missing full stop

noman-land committed:
whisper: Fix spelling and grammar in error (#15009)

* whisper: Fix spelling and grammar in error

* whisper: Fix grammar in comments

karalabe committed:
Merge pull request #14983 from karalabe/metropolis-revert

core/vm: implement REVERT metropolis opcode

george-angel committed:
Dockerfile: use alpine:3.6, clean up apk incovation (#15006)

karalabe committed:
vendor: update USB HID lib (fix macOS crash)

karalabe committed:
Merge pull request #15017 from karalabe/usbhid-macos-thread-fix

vendor: update USB HID lib (fix macOS crash)

Bo-Ye committed:
metrics: change MetricsEnabledFlag to be const

tizhou86 committed:
core/vm: fix typo in method documentation (#15019)

Signed-off-by: Ti Zhou <tizhou1986@gmail.com>

karalabe committed:
tests: enable Byzantium state tests for CI

karalabe committed:
Merge pull request #15022 from karalabe/enable-byzantium-state-tests

tests: enable Byzantium state tests for CI

karalabe committed:
tests: pull in new test suite, enable most block tests

karalabe committed:
core/types: reject Metro receipts with > 0x01 status bytes

karalabe committed:
Merge pull request #15018 from Bo-Ye/master

metrics: change MetricsEnabledFlag to be const

karalabe committed:
Merge pull request #15000 from fjl/node-flock

node: fix instance dir locking and improve error message

holiman committed:
cmd/evm, core/vm, internal/ethapi: Show error when exiting (#14985)

* cmd/evm, core/vm, internal/ethapi: Add 'err' to tracer interface CaptureEnd

* cmd/evm: fix nullpointer when there is no error

karalabe committed:
Merge pull request #15014 from rjl493456442/metropolis-eip658

core: add status as a consensus field in receipt

karalabe committed:
rlp: fix decoding long strings into RawValue types

karalabe committed:
Merge pull request #15029 from karalabe/rlp-raw-decode

rlp: fix decoding long strings into RawValue types

karalabe committed:
core/state: revert metro suicide map addition (#15024)

Commit:
core/types: fix create indicator in Transaction.String (#15025)

bts committed:
Dynamic raft membership support (#130)

patrickmn committed:
private: Distinguish between privateFor being absent vs. an empty array to allow self-sending (#165)

joelburget committed:
Small spelling / grammar tweaks to design.md (#164)

joelburget committed:
README.md: Add a few video links

gsalgado committed:
Add docs/new-block-code-walkthrough.md walkthrough for the creation of a new block (#140)

coeniebeyers committed:
Adding information regarding enode order in static-nodes.json (#174)

stevenroose committed:
Use latest block when pending is indicated with RAFT (#181)

Commit:
Fix the dash in the bootnode commands (#182)

patrickmn committed:
README.md: Add ZSL links

patrickmn committed:
README.md: Add ZSL links (2)

joelburget committed:
Don't raise future block errors in Quorum mode.

bts committed:
Fix async tx submission

rojotek committed:
Added a format directive to the fatalf call in raft/snapshot.bytesToSnapshot() (#197)

Reformatted import statements

joelburget committed:
Make IsQuorum the default configuration
